streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyJoinIntegrationTest.java

```diff
@@ -27,673 +25,314 @@
-            else if (two != null)
-                result = two;
-            else if (three != null)
-                result = three;
-            else
-                throw new RuntimeException("Cannot find key " + expected.getKey() + " in any of the state stores");
-            assertEquals(expected.getValue(), result);
-        }
-
-        //Merge all the iterators together to ensure that their sum equals the total set of expected elements.
-        final KeyValueIterator<Integer, String> allOne = myJoinStoreOne.all();
-        final KeyValueIterator<Integer, String> allTwo = myJoinStoreTwo.all();
-        final KeyValueIterator<Integer, String> allThree = myJoinStoreThree.all();
-
-        final List<KeyValue<Integer, String>> all = new LinkedList<>();
-
-        while (allOne.hasNext()) {
-            all.add(allOne.next());
-        }
-        while (allTwo.hasNext()) {
-            all.add(allTwo.next());
-        }
-        while (allThree.hasNext()) {
-            all.add(allThree.next());
-        }
-        allOne.close();
-        allTwo.close();
-        allThree.close();
-
-        for (final KeyValue<Integer, String> elem : all) {
-            assertTrue(expectedResult.contains(elem));
-        }
+    private static Map<Integer, String> asMap(final KeyValueStore<Integer, String> store) {
+        final HashMap<Integer, String> result = new HashMap<>();
+        store.all().forEachRemaining(kv -> result.put(kv.key, kv.value));
+        return result;
     }
 
-    private KafkaStreams prepareTopology(final String queryableStoreName, final boolean leftJoin) {
+    private static Topology getTopology(final Properties streamsConfig,
+                                        final String queryableStoreName,
+                                        final boolean leftJoin) {
         final StreamsBuilder builder = new StreamsBuilder();
 
         final KTable<Integer, Float> left = builder.table(LEFT_TABLE, Consumed.with(Serdes.Integer(), Serdes.Float()));
         final KTable<String, Long> right = builder.table(RIGHT_TABLE, Consumed.with(Serdes.String(), Serdes.Long()));
 
-        final Materialized<Integer, String, KeyValueStore<Bytes, byte[]>> materialized;
-        if (queryableStoreName != null) {
-            materialized = Materialized.<Integer, String, KeyValueStore<Bytes, byte[]>>as(queryableStoreName)
-                    .withKeySerde(Serdes.Integer())
-                    .withValueSerde(Serdes.String())
-                    .withCachingDisabled();
-        } else {
-            throw new RuntimeException("Current implementation of join on foreign key requires a materialized store");
-        }
+        final Materialized<Integer, String, KeyValueStore<Bytes, byte[]>> materialized =
+            Materialized.<Integer, String>as(Stores.inMemoryKeyValueStore(queryableStoreName))
+                .withKeySerde(Serdes.Integer())
+                .withValueSerde(Serdes.String())
+                .withCachingDisabled();
```

Comment:
Should we test with caching (to make sure we execute the caching code path)? Note that TDD flushes after each `pipeInput()` anyway.

Reply:
yeah, I think it's an artifact. I'll remove it.

Reply:
Huh. It looks like removing this results in not seeing all the tombstones we expect. I'll leave it in.

Reply:
That is weird. Can you investigate? It sound like it exposes a bug.

Reply:
Ok, I'll look into it.

Reply:
If you're ok with it, I'll continue this investigation in a separate PR. It would be good to at least get this test improvement merged before figuring out why that cache is misbehaving.

Reply:
Found in org.apache.kafka.streams.state.internals.CachingKeyValueStore#putAndMaybeForward:

```java
// this is an optimization: if this key did not exist in underlying store and also not in the cache,
// we can skip flushing to downstream as well as writing to underlying store
```

So, it actually is expected that in situations where the join result is actually materialized, we would not see duplicate tombstones (which some of the tests expect).

With this understanding in place, I'm going to leave this line here for now. I'm preparing a follow-on PR to make the `Materialized` parameter optional (and queriability optional even if you specify it). For that PR, I'll have to make specific assertions about this behavior anyway, so we'll have good test coverage there.

Reply:
Thanks!